File: packages/api/src/models/alert.ts
import { ObjectId } from 'mongodb';

/**
 * Alert rule schema
 * Requirements: 10.1, 10.3, 10.4
 */

export type AlertOperator = '>' | '<' | '==' | '!=' | 'change_pct' | 'change_abs';
export type AlertSeverity = 'info' | 'warning' | 'critical';
export type NotificationChannel = 'email' | 'webhook' | 'in_app' | 'sms';

export interface AlertCondition {
  metric: string;
  operator: AlertOperator;
  threshold: number;
  duration?: number;
  cooldown?: number;
}

export interface WebhookConfig {
  url: string;
  method: 'POST' | 'GET';
  headers?: Record<string, string>;
  body?: Record<string, any>;
}

export interface EmailConfig {
  to: string[];
  cc?: string[];
  subject?: string;
  template?: string;
}

export interface SmsConfig {
  phoneNumbers: string[];
  message?: string;
}

export interface NotificationChannelConfig {
  type: NotificationChannel;
  enabled: boolean;
  webhook?: WebhookConfig;
  email?: EmailConfig;
  sms?: SmsConfig;
}

export interface AlertRule {
  _id?: ObjectId;
  alertRuleId: string;
  userId: string;
  name: string;
  description?: string;
  condition: AlertCondition;
  channels: NotificationChannelConfig[];
  enabled: boolean;
  tags?: string[];
  createdAt: Date;
  updatedAt: Date;
  lastTriggeredAt?: Date;
  triggerCount: number;
}

export interface AlertHistory {
  _id?: ObjectId;
  alertHistoryId: string;
  alertRuleId: string;
  userId: string;
  timestamp: Date;
  metric: string;
  currentValue: number;
  threshold: number;
  severity: AlertSeverity;
  context: Record<string, any>;
  suggestedActions?: string[];
  acknowledged: boolean;
  acknowledgedAt?: Date;
  acknowledgedBy?: string;
  notes?: string;
}

export const ALERT_RULE_COLLECTION = 'alert_rules';
export const ALERT_HISTORY_COLLECTION = 'alert_history';

/**
 * MongoDB indexes for alert rule collection
 */
export const ALERT_RULE_INDEXES = [
  {
    key: { alertRuleId: 1 },
    unique: true,
    name: 'idx_alertRuleId',
  },
  {
    key: { userId: 1 },
    name: 'idx_userId',
  },
  {
    key: { userId: 1, enabled: 1 },
    name: 'idx_userId_enabled',
  },
  {
    key: { 'condition.metric': 1 },
    name: 'idx_metric',
  },
  {
    key: { tags: 1 },
    name: 'idx_tags',
  },
];

/**
 * MongoDB indexes for alert history collection
 */
export const ALERT_HISTORY_INDEXES = [
  {
    key: { alertHistoryId: 1 },
    unique: true,
    name: 'idx_alertHistoryId',
  },
  {
    key: { alertRuleId: 1 },
    name: 'idx_alertRuleId',
  },
  {
    key: { userId: 1 },
    name: 'idx_userId',
  },
  {
    key: { userId: 1, timestamp: -1 },
    name: 'idx_userId_timestamp',
  },
  {
    key: { timestamp: -1 },
    name: 'idx_timestamp',
  },
  {
    key: { acknowledged: 1 },
    name: 'idx_acknowledged',
  },
];

File: packages/agents/src/models/alert.py
"""Alert rule models for MongoDB"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class AlertCondition(BaseModel):
    """Alert condition specification"""
    metric: str
    operator: str
    threshold: float
    duration: Optional[int] = None
    cooldown: Optional[int] = None


class WebhookConfig(BaseModel):
    """Webhook notification configuration"""
    url: str
    method: str = "POST"
    headers: Optional[Dict[str, str]] = None
    body: Optional[Dict[str, Any]] = None


class EmailConfig(BaseModel):
    """Email notification configuration"""
    to: List[str]
    cc: Optional[List[str]] = None
    subject: Optional[str] = None
    template: Optional[str] = None


class SmsConfig(BaseModel):
    """SMS notification configuration"""
    phone_numbers: List[str]
    message: Optional[str] = None


class NotificationChannelConfig(BaseModel):
    """Notification channel configuration"""
    type: str
    enabled: bool = True
    webhook: Optional[WebhookConfig] = None
    email: Optional[EmailConfig] = None
    sms: Optional[SmsConfig] = None


class AlertRule(BaseModel):
    """Alert rule document"""
    alert_rule_id: str
    user_id: str
    name: str
    description: Optional[str] = None
    condition: AlertCondition
    channels: List[NotificationChannelConfig]
    enabled: bool = True
    tags: Optional[List[str]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    last_triggered_at: Optional[datetime] = None
    trigger_count: int = 0

    class Config:
        """Pydantic configuration"""
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class AlertHistory(BaseModel):
    """Alert history document"""
    alert_history_id: str
    alert_rule_id: str
    user_id: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    metric: str
    current_value: float
    threshold: float
    severity: str
    context: Dict[str, Any] = Field(default_factory=dict)
    suggested_actions: Optional[List[str]] = None
    acknowledged: bool = False
    acknowledged_at: Optional[datetime] = None
    acknowledged_by: Optional[str] = None
    notes: Optional[str] = None

    class Config:
        """Pydantic configuration"""
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


# Collection names
ALERT_RULE_COLLECTION = "alert_rules"
ALERT_HISTORY_COLLECTION = "alert_history"

# Index specifications
ALERT_RULE_INDEXES = [
    {"keys": [("alert_rule_id", 1)], "unique": True, "name": "idx_alertRuleId"},
    {"keys": [("user_id", 1)], "name": "idx_userId"},
    {"keys": [("user_id", 1), ("enabled", 1)], "name": "idx_userId_enabled"},
    {"keys": [("condition.metric", 1)], "name": "idx_metric"},
    {"keys": [("tags", 1)], "name": "idx_tags"},
]

ALERT_HISTORY_INDEXES = [
    {"keys": [("alert_history_id", 1)], "unique": True, "name": "idx_alertHistoryId"},
    {"keys": [("alert_rule_id", 1)], "name": "idx_alertRuleId"},
    {"keys": [("user_id", 1)], "name": "idx_userId"},
    {"keys": [("user_id", 1), ("timestamp", -1)], "name": "idx_userId_timestamp"},
    {"keys": [("timestamp", -1)], "name": "idx_timestamp"},
    {"keys": [("acknowledged", 1)], "name": "idx_acknowledged"},
]
